### PR TITLE
Vala language support

### PR DESCRIPTION
**Pull request by matbme**
Support for the [Vala](https://wiki.gnome.org/Projects/Vala) programming language.

I *think* I included everything that's necessary, but if not let me know! :)

### CODE REVIEW

File: ftdetect/vala.vim [hunk 1]
```diff
@@ -0,0 +1 @@
+au BufRead,BufNewFile *.vala set filetype=vala
```
Comment (clason) [inline ftdetect/vala.vim:1]: We are trying to get rid of the ftdetects; these should go into upstream. (The rest can still be merged.)

Comment (matbme) [reply ftdetect/vala.vim]: Removed

File: queries/vala/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,387 @@
+; Pointers
+
+(address_of_identifier
+  "&" @symbol
+  (_)*
```
Comment (theHamsta) [inline queries/vala/highlights.scm:5]: this has no effect

File: queries/vala/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+; Booleans
+
+(true) @boolean
+(false) @boolean
+
+; Operators
+
+(binary_expression
+  (_)
```
Comment (theHamsta) [inline queries/vala/highlights.scm:52]: you don't have to name every child of an expression.

File: queries/vala/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+  ]
+)
+
+(function_call
+  identifier: [
+  	(identifier) @method
+    (camel_cased_identifier) @type
+    (generic_identifier (_) @type) 
+    (namespaced_identifier
+    	left: (_)
```
Comment (theHamsta) [inline queries/vala/highlights.scm:119]: if you don't need `left` for you query, there's also no need to add it to this file

File: queries/vala/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    "?" @symbol
+)
+
+"typeof" @keyword
+
+"is" @keyword
+
+(is_type_expression
+  (_)
+  "is" @keyword
```
Comment (theHamsta) [inline queries/vala/highlights.scm:146]: but the previous, query hightlights `"is"` already always.

File: queries/vala/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+(string_template_variable) @variable
+
+(string_template_expression) @variable
+
+; Assignment and declaration
+
+(new_instance
+  [
+    (
+      "new" @keyword
```
Comment (theHamsta) [inline queries/vala/highlights.scm:236]: I assume that `"new"` can only be used this way in the grammar. Wouldn't a simple list of all the keywords be more maintainable?

File: queries/vala/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+  ))
+  (block)
+)
+
+; Enum
+
+"enum" @keyword
+(enum_declaration
+    (camel_cased_identifier) @type
+    (_)*
```
Comment (theHamsta) [inline queries/vala/highlights.scm:284]: This looks, all in all good, but I have the feeling that you're just repeating every rule of the grammar in the queries. They could be simplified.

File: queries/vala/highlights.scm (hunk 2 above)
Comment (matbme) [reply queries/vala/highlights.scm:5]: Difference

The intention is to highlight the address symbol differently from the and operator

**Comment by matbme** [comment]
@theHamsta thanks for the feedback! I pushed a new commit that removes unnecessary entries in highlights.scm

File: queries/vala/highlights.scm (hunk 2 above)
Comment (theHamsta) [reply queries/vala/highlights.scm]: but `(_)*` is not needed for that. `(_)*` mean zero or more anything which does not add any information to the query.

**Review by theHamsta — APPROVED** [review]

**Comment by matbme** [comment]
Also, Bram pushed a [patch](https://github.com/vim/vim/commit/97c554d5149c2aa4a43d689c59563e77277265d4) adding Vala filetype support. Now we just wait for Neovim to merge?

File: queries/vala/highlights.scm [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+  "return" 
+  "yield"
+  "break"
+] @keyword.return
+
+[
+  (null)
+  (modifier)
+  "typeof"
+  "is"
```
Comment (theHamsta) [inline queries/vala/highlights.scm:37]: maybe `"is"` and `typeof` are `@keyword.operator`

File: queries/vala/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+; Reserved keywords
+
+[
+  "return" 
+  "yield"
+  "break"
+] @keyword.return
+
+[
+  (null)
```
Comment (theHamsta) [inline queries/vala/highlights.scm:34]: `(null)` is maybe `@constant.builtin`

File: queries/vala/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+; Enum
+
+(enum_declaration
+    (camel_cased_identifier) @type
+)
+
+; Loop
+
+(foreach_statement
+  loop_item: (identifier) @variable
```
Comment (theHamsta) [inline queries/vala/highlights.scm:240]: you have a query at the end that highlights every `identifier` as `@variable`

File: queries/vala/highlights.scm [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+    (identifier) @parameter
+    (camel_cased_identifier) @type
+    (uppercased_identifier) @constant
+  ]
+)
+
+; Variable
+
+(identifier) @variable
+(camel_cased_identifier) @variable
```
Comment (theHamsta) [inline queries/vala/highlights.scm:284]: usually those two queries are at the beginning of the query files so that other queries can overwrite

**Comment by clason** [comment]
> Also, Bram pushed a [patch](https://github.com/vim/vim/commit/97c554d5149c2aa4a43d689c59563e77277265d4) adding Vala filetype support. Now we just wait for Neovim to merge?

Yes. Won't have to wait long.

**Comment by clason** [comment]
And it's merged.

File: queries/vala/highlights.scm (hunk 11 above)
Comment (matbme) [reply queries/vala/highlights.scm:284]: For some reason placing them at the top yields the opposite result. Elements inside rules like function_call and code_attribute get overwritten

**Review by matbme — COMMENTED** [review]
@theHamsta changed the highlights you suggested. If everything else is ready, I believe it is ready to merge

**Comment by theHamsta** [comment]
Nice work!

File: queries/vala/highlights.scm [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    "&&"
+    "||"
+    "&"
+    "|"
+    "^"
+    "~"
+    "|="
+    "&="
+    "^="
+    "??"
```
Comment (theHamsta) [inline queries/vala/highlights.scm:127]: I think `"="` is missing

File: queries/vala/highlights.scm [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+; Misc
+
+(number) @number
+
+[
+  "{"
+  "}"
+  "("
+  ")"
+] @punctuation.bracket
```
Comment (theHamsta) [inline queries/vala/highlights.scm:31]: I think `"["` `"]"` is missing. Could you add those two details please? Just tried out your vala highlighting for the first time.

Comment (theHamsta) [reply queries/vala/highlights.scm:31]: `void open (File[] files, string hint) {` array_identifier also does not get highlighted as a type. But that can be a follow up

Comment (matbme) [reply queries/vala/highlights.scm:33]: My bad, must have removed them somewhere during simplification

Comment (matbme) [reply queries/vala/highlights.scm:33]: > `void open (File[] files, string hint) {` array_identifier also does not get highlighted as a type. But that can be a follow up

image

I believe it was, my last commit seems to fix it.

File: queries/vala/highlights.scm (hunk 12 above)
Comment (theHamsta) [reply queries/vala/highlights.scm:129]: ":" could be added to the `punctuation.delimiter` section

Comment (theHamsta) [reply queries/vala/highlights.scm:129]: ":" could be added to the `punctuation.delimiter` section

**Pull request merged**